### PR TITLE
Handle HTML error responses from APIC

### PR DESCRIPTION
**Pull request by tycha-d-1-3**
Issue: Client code sometimes returns with an error - invalid character '<' looking for beginning of value

Root Cause: There are certain scenarios where the response is an nginx html response rather than a JSON error message
    Examples:
        - Page does not exist
        - nginx on APIC is too busy
        - etc.

Fix: When the JSON Error parsing fails, fallback to HTML parsing and do a best effort gather of the error response from the HTML page.

Test Response when trying to perform a GET against the visore.html page:
```
API URL: https://dev8-ifc1.insieme.local/visore.html, Verify that you are connecting to an APIC.
HTTP response status: 200 OK
Message: You need to enable JavaScript to run this app.
```

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -541,9 +542,14 @@ func (c *Client) Do(req *http.Request) (*container.Container, *http.Response, er
...
-				log.Printf("[ERROR] Error occured while json parsing %+v", err)
+				log.Printf("[ERROR] Error occured while json parsing: %+v", err)
+
+				// If nginx is too busy or the page is not found, APIC's nginx will response with an HTML doc instead of a JSON Response.
+				// In those cases, parse the HTML response for the message and return that to the user
+				htmlErr := c.checkHtmlResp(bodyStr)
+				log.Printf("[ERROR] Error occured while json parsing: %s", htmlErr.Error())
 				log.Printf("[DEBUG] Exit from Do method")
-				return nil, resp, errors.New(fmt.Sprintf("Failed to parse JSON response from %s. Verify that you are connecting to an APIC.\nHTTP response status: %s\nMessage: %+v", req.URL.String(), resp.Status, err))
+				return nil, resp, errors.New(fmt.Sprintf("API URL: %s, Verify that you are connecting to an APIC.\nHTTP response status: %s\nMessage: %s", req.URL.String(), resp.Status, htmlErr))
```
Comment (lhercot) [inline client/client.go:552]: Why change the log error message being returned?

File: client/client.go [hunk 2]
```diff
@@ -552,9 +558,14 @@ func (c *Client) Do(req *http.Request) (*container.Container, *http.Response, er
 			if ok := c.backoff(attempts); !ok {
 				obj, err := container.ParseJSON(bodyBytes)
 				if err != nil {
-					log.Printf("[ERROR] Error occured while json parsing %+v with HTTP StatusCode 405, 500-504", err)
+					log.Printf("[ERROR] Error occured while json parsing: %+v", err)
```
Comment (lhercot) [inline client/client.go:561]: please keep the "with HTTP StatusCode 405, 500-504" part so we differentiate the log message from L544

File: client/client.go [hunk 3]
```diff
@@ -552,9 +558,14 @@ func (c *Client) Do(req *http.Request) (*container.Container, *http.Response, er
...
+
+					// If nginx is too busy or the page is not found, APIC's nginx will response with an HTML doc instead of a JSON Response.
+					// In those cases, parse the HTML response for the message and return that to the user
+					htmlErr := c.checkHtmlResp(bodyStr)
+					log.Printf("[ERROR] Error occured while json parsing: %s", htmlErr.Error())
 					log.Printf("[DEBUG] Exit from Do method")
-					return nil, resp, errors.New(fmt.Sprintf("Failed to parse JSON response from %s. Verify that you are connecting to an APIC.\nHTTP response status: %s\nMessage: %+v", req.URL.String(), resp.Status, err))
+					return nil, resp, errors.New(fmt.Sprintf("API URL: %s, Verify that you are connecting to an APIC.\nHTTP response status: %s\nMessage: %s", req.URL.String(), resp.Status, htmlErr))
```
Comment (lhercot) [inline client/client.go:568]: Why change the log error message being returned?

**Review by lhercot — CHANGES_REQUESTED** [review]

File: client/client.go (hunk 1 above)
Comment (tycha-d-1-3) [reply client/client.go:552]: the APIC is returning either an JSON body or an HTML body.  if you return this error message, its not actually conveying the error message that is being returned from APIC.  If JSON Parsing fails, that means APIC is returning an HTML Body - so what i'm doing here is extracting the message from the HTML body and returning that as the actual error message.

File: client/client.go (hunk 3 above)
Comment (tycha-d-1-3) [reply client/client.go:568]: the APIC is returning either an JSON body or an HTML body.  if you return this error message, its not actually conveying the error message that is being returned from APIC.  If JSON Parsing fails, that means APIC is returning an HTML Body - so what i'm doing here is extracting the message from the HTML body and returning that as the actual error message.

File: client/client.go (hunk 2 above)
Comment (tycha-d-1-3) [reply client/client.go:561]: Sure - I can I can keep this part.

File: client/client.go (hunk 1 above)
Comment (lhercot) [reply client/client.go:552]: I am only talking about keeping the start of the log message "Failed to parse JSON response from %s." we can keep the extracted HTML body part. Just not sure why we don't keep the start of the log message as "API URL: %s" does not really tell us what is happening either.

File: client/client.go (hunk 3 above)
Comment (lhercot) [reply client/client.go:568]: I am only talking about keeping the start of the log message "Failed to parse JSON response from %s." we can keep the extracted HTML body part. Just not sure why we don't keep the start of the log message as "API URL: %s" does not really tell us what is happening either.

**Comment by lhercot** [comment]
Thanks for the contribution!

**Pull request merged**